### ISSUE TITLE
Alternative relay server (STUN/TURN/ICE/Tor)

### ISSUE THREAD

**Issue by ariddell** [warner/magic-wormhole#6]
I've been reading a bit lately about ICE for NAT traversal (https://en.wikipedia.org/wiki/Interactive_Connectivity_Establishment). Is there any interest in supporting this as the first option instead of the (comparatively) heavy-weight relay service?

Or perhaps you're more interested in routing via tor? That would, of course, solve the NAT problem as well.

**Comment by warner** [comment]
I'd love to do both, and yeah using ICE as the primary, with the transit relay as a fallback. I'm planning to add a `--tor` option that both turns on the hidden-service transit and does all the rendezvous-server communication over Tor.

If I have my NAT-bypass techniques straight, I think ICE is the one where you maintain a TCP connection outbound (from a specific local TCP source port), and then that enables you to receive inbound TCP connections (using that same port), right?. Basically I want to do whatever clever thing IPFS is doing, which I think looks like that. So I think we can use the rendezvous server as the "what is my source IP / source port" server, and the `EventSource` channel as the held-open TCP connection.

**Comment by david415** [comment]
IPFS uses ICE.

**Comment by david415** [comment]
You might checkout meejah's carml utility:
https://github.com/meejah/carml

It has a pastebin and copybin commands... which now work with stealth authenticated hidden services:
https://github.com/meejah/carml/blob/master/carml/command/pastebin.py
https://github.com/meejah/carml/blob/master/carml/command/copybin.py

Authenticated hidden services require the client to not only know the onion address but the hidden service cookie as well.

**Comment by ariddell** [comment]
Thanks. It seems like ICE might be the easiest place to start. I'll look around for a Python implementation.

**Comment by warner** [comment]
I read a couple of wikipedia pages and RFCs last night. The division of responsibilities seems to be:

* STUN: ask `ifconfig` and a server "what is my external IP/port" (for UDP)
* ICE: try a bunch of connection hints until something works
* TURN: relay data through an external (public) server

The RFCs make it clear that this is all somewhat SIP- and VOIP- centric, so it all runs over UDP.

`wormhole.blocking.transit` is currently doing something morally equivalent to ICE (trying various host/port targets until one of them negotiates successfully), and the directly-connected/local-interface/ifconfig portion of STUN.

On one hand, it might be better to make wormhole's Transit code implement the actual STUN/ICE/TURN protocols, so we can take advantage of the existing work that's gone into them. Since we're not running SIP (and I don't think I want us too.. it sounds more complicated than it's worth), we either need a polite flow-controlled UDP-based stream transport (maybe uTP?), or we need a form of STUN/ICE/etc that works for TCP. I've heard of "STUNT" (STUN for TCP), but in my quick read of the specs it all seemed limited to UDP.

We also need to look at how ICE decides that it's successfully made a connection: wormhole.transit does a cryptographic handshake to confirm that we're talking to the correct endpoint (and not just some random server that happened to be listening on the same port), and ICE might not. The RFC I skimmed suggested that a MAC-based authentication might be used, but it's quite possible that the ICE security model is either glued deeply to TLS, or insufficiently strong for our tastes. WebRTC (see below) is glued to DTLS, which might be good enough for connection-establishment purposes, and we can always run our own encryption inside the DTLS tunnel.

On the other hand, if the protocols are too domain-specific, it might not be worth the bother. Or rather it might be worth more bother to achieve the same goals with something custom.

We should definitely look at what IPFS does, because they've got a similar problem space. I'm hoping for a solution that doesn't require too much C code. IPFS can be kinda casual about dragging in dependencies because it's written in Go, but for each C library we depend upon in Python, we lose some deployability to folks without compilers.

Also we should be looking towards the web: to allow browser-based implementations to interoperate with CLI tools, we may be limited to what a browser can do, which is WebSocket (now) and WebRTC (now/soonish). I haven't found any up-to-date pure-python WebSocket implementations, but it sounds feasible, and I think there are run-a-helper-in-a-subprocess ways to do it. Other languages have more support.

WebRTC is SIP-based (using STUN/ICE/etc), and a lot more complex (most of the libraries focus on the H.264 codec stuff, which we obviously don't care about (yet)), so I'm not as optimistic about finding easy libraries. *But*, WebRTC is supposed to have all the p2p stuff built-in. So in an ideal world, `pip install magic-wormhole` would give you a WebRTC-based transport directly to the other client.

A few pointers:

* WebSockets
 * https://github.com/liris/websocket-client (blocking)
 * https://github.com/tuxlinuxien/TwistedWebsocket
 * https://github.com/aaugustin/websockets (py3 asyncio)
 * https://github.com/Pithikos/python-websocket-server
 * https://github.com/gregmuellegger/django-websocket
* WebRTC
 * https://github.com/theintencity/py-webrtc
 * https://github.com/EricssonResearch/OpenWebRTC
* ICE
 * https://en.wikipedia.org/wiki/Interactive_Connectivity_Establishment 
 * RFC5245 https://tools.ietf.org/html/rfc5245
* STUN
 * https://en.wikipedia.org/wiki/Session_Traversal_Utilities_for_NAT
 * RFC5389 https://tools.ietf.org/html/rfc5389
* TURN
 * https://en.wikipedia.org/wiki/Traversal_Using_Relays_around_NAT 
 * RFC5766 https://tools.ietf.org/html/rfc5766

**Comment by ariddell** [comment]
I was also looking at this: https://github.com/getlantern/natty, which happens to include a list of public STUN servers: https://github.com/getlantern/natty/blob/master/STUNS.md

It seems like the pure-python requirement (a good one!) might be the killer for using WebRTC in the short term. Maybe it would be easier just to start with STUN if we're sticking to pure python? I wish I knew how useful STUN alone is -- e.g., does it succeed more than 50% of the time with users in NAT-heavy places like China? What's its failure rate in the US?

**Comment by ariddell** [comment]
Two more items:
- Someone (who seems knowledgeable) suggests STUN works in 70% of cases on laptops and desktops. https://stackoverflow.com/questions/23655243/nat-traversal-probability-of-success-using-stun
- There's Python code to use the WebSocket protocol over arbitrary streams with Twisted (I think that's a reasonable description) https://github.com/crossbario/autobahn-python/tree/master/examples/twisted/websocket/wrapping

**Comment by david415** [comment]
Dear @ariddell It's the NAT device or lack thereof (the network topology) that would allow STUN to work or not... It should not depend on the setup of the desktop/laptop device itself. Consider that NAT violates the end-to-end principle; in that case an intermediary is performing work on your behalf.

@warner if the magic-wormhole relay server listened on an onion service then you could hardcode that onion address into the source... and it should work similarly to how it works now. One of the advantage of course is that the relay server could change hosting providers but it's onion wouldn't change.

**Comment by ariddell** [comment]
@davis415 An onion service sounds best for the reasons you provide. What's the strategy for getting an onion service working here for the client? Doesn't tor need to be bundled with magic-wormhole in order to use something like txtorcon? Or is the idea that magic-wormhole could detect if tor was running on the system and, if so, use it?

**Comment by warner** [comment]
I just learned about [RFC6062](http://tools.ietf.org/html/rfc6062), which is TURN for TCP (almost identical to what my transit relay is currently doing). And Twilio's [Network Traversal Service](https://www.twilio.com/stun-turn) claims RFC-6062 compatibility. I'm still looking for Twisted (or just python) implementations of clients for any of these, and I'm still trying to figure out if there's a name/RFC for STUN-but-with-TCP.

Twilio offers free STUN service (the "what is my address" server), and pay-by-the-GB TURN service. If the code is feasible, I might just use them for the relay, especially since that design would probably let other magic-wormhole library users (e.g. apps) pick their own relay service.

**Comment by paride** [comment]
It seems to me that STUN supports TCP (and TLS over TCP):

https://tools.ietf.org/html/rfc5389#section-7.2.2

It would be great if wormhole supported standard STUN/ICE/TURN.

**Comment by kronicd** [comment]
Agree, it would help eliminate the need for the orchestration server to
relay data so much.

**Comment by meejah** [comment]
@aridell often people have system Tor instances running; txtorcon can work with those or start its own tor client. I personally wouldn't bundle tor, but direct people to torproject.org to download their own.

**Comment by kronicd** [comment]
It might be worth taking a look at katz: https://github.com/gnubert/katz/

As long as the public IP & ports of both ends are known it sets up very reliable direct connections with both ends behind NAT. It is UDP only as far as I know though.

**Comment by dwoz** [comment]
Just for posterity, I was playing around with a toy implementation of hole punching when I opend #227. Here is my example showing [tcp/udp hole punching in pure python](https://github.com/dwoz/python-nat-hole-punching).

**Comment by dwoz** [comment]
Here are two examples of webrtc in python:

[https://github.com/saghul/CallRoulette](https://github.com/saghul/CallRoulette)
[https://github.com/saghul/CallRoulette](https://github.com/pubnub/SimpleRTC)

**Comment by meejah** [comment]
Autobahn (websocket library) is pure-python enough to work with PyPy...

**Comment by dwoz** [comment]
I've learned a bit more about this stuff in the past months. If a goal is to support WebRTC as one of the transports. Then we're talking support for ICE/(STUN/TURN) which creates and RTC connection. The RTC connection can be TCP or UDP, I think supporting both is ideal. With WebRTC, you still need a relay mechanism for the SDP, so essentially a relay server. The SDP is a block of text that contains information about ICE candidates and the protocols you want to use for video/audio/data channels. I think but am not 100% that one could leverage the SDP (extension mechanisms) to transmit the PAKE, Key Agreement. The relay mechanism is outside of the scope of WebRTC itself, people use both websockets and plain http(s) for this (it can be anything though, even email).

Another link for reference, looks like it's got all the major pieces but I'm not sure about the license situation:

https://github.com/theintencity/rtclite

**Comment by joshtriplett** [comment]
libnice (mentioned above) does seem like it'd provide all the support needed to establish a direct connection through NAT.

**Comment by bakkot** [comment]
https://github.com/aiortc/aiortc appears to be a Python library for WebRTC.

**Comment by piegamesde** [comment]
At the moment, I am experimenting with using UDP-based reliable transport protocols like UDT as replacement for TCP in the relay. This gives us:

- The ability to open an UDP connection and use vanilla STUN services to get the public IP, translate the socket into a reliable UDT connection for transit
- Apparently, UDT and other newer protocols solve a few problems like flow control in a "better" way than TCP, so it's nice to have in general.

Downside: new relay negotiation hints must be established.

**Comment by marek22k** [comment]
Hi, if you are planning to add Tor support, maybe some for [I2P](https://geti2p.net/en/) too?

**Comment by meejah** [comment]
@marek22k there is already Tor support for the client-type connections, but you are probably getting at something more like https://github.com/magic-wormhole/magic-wormhole/issues/8 I think?

**Comment by marek22k** [comment]
Yes, but just the same with the I2P network. I2P provides both [SOCKS support](https://geti2p.net/en/docs/api/socks) and an [API](https://geti2p.net/en/docs/api/samv3) to create "hidden services".

**Comment by HuakunShen** [comment]
WebRTC seems to be very slow. My implementation in rust to transfer a file is ~10MBps. While magic wormhole gives full speed ~110MBps under a 1000mbps local network.
